app/models/shipit/deploy_spec/npm_discovery.rb

```diff
@@ -79,10 +85,84 @@ def package_version
         JSON.parse(package_json.read)['version']
       end
 
+      def publish_config
+        JSON.parse(package_json.read)['publishConfig']
+      end
+
+      def publish_config_access
+        config = publish_config
+
+        # default to private deploy when we enforce a publishConfig
+        if enforce_publish_config?
+          return PRIVATE if config.blank?
+          config['access'] || PRIVATE
+        end
+
+        return PUBLIC if config.blank?
+        config['access'] || PUBLIC
+      end
+
+      def package_name
+        JSON.parse(package_json.read)['name']
+      end
+
+      def scoped_package?
+        package_name.start_with?('@shopify')
+      end
+
+      def enforce_publish_config?
+        enforce = Shipit.enforce_publish_config
+        return false if enforce.nil? || enforce.to_s == "0"
+        true
+      end
+
+      def valid_publish_config?
+        return true unless enforce_publish_config?
+        return false if Shipit.private_npm_registry.nil?
+        return false if publish_config.blank?
+        return true if publish_config_access == PUBLIC
+
+        valid_publish_config_access? && private_scoped_package?
+      end
+
+      def valid_publish_config_access?
+        VALID_ACCESS.include?(publish_config_access)
+      end
+
+      # ensure private packages are scoped
+      def private_scoped_package?
+        publish_config_access == PRIVATE && scoped_package?
+      end
+
+      def local_npmrc
+        file(".npmrc")
+      end
+
+      def npmrc_contents(registry)
+        "always-auth=true\n#{registry}"
+      end
+
+      def registry
+        prefix = scoped_package? ? "@shopify:registry" : "registry"
```

Comment:
Sorry but hardcoding `@shopify:registry` is not acceptable, because Shopify is not the sole user of Shipit.

I know very little about NPM & such, so if the way we're doing package publishing is specific to our organization, then this code shouldn't be inside `shipit-engine`, you can implement all that inside our private installation of Shipit.

If it's the way most people do package publishing, then go ahead but you'll have to make this org name configurable somehow.

Reply:
Oh yes, good point. I'll pull out the private scope into shipit as well.